src/Storage/Notmuch.hs

```diff
@@ -40,6 +40,22 @@ getMessages s settings =
               mails <- liftIO $ mapM messageToMail msgs
               pure $ Vec.fromList mails
 
+-- | creates a vector of parsed mails which are tips of each thread
+--
+getTopMessages
+  :: (HasThreads Query, MonadError Error m, MonadIO m)
+  => T.Text
+  -> NotmuchSettings FilePath
+  -> m (Vec.Vector NotmuchMail)
+getTopMessages s settings =
+  bracketT (databaseOpenReadOnly (view nmDatabase settings)) databaseDestroy go
+  where
+    go db = do
+        tds <- query db (FreeForm $ T.unpack s) >>= threads >>= mapM threadToplevelMessages
+        let flattened = foldl (++) [] tds
```

Comment:
Use `(<>)` instead of `(++)`, and I think you probably want `foldr` instead of `foldl`.

Or because it is `[[a]]` you could replace `flattened` with `Control.Monad.join`.

Reply:
Forgetting this all the time, but the code has changed since then.